pilot/pkg/networking/core/v1alpha3/tls.go

```diff
@@ -260,3 +249,31 @@ func buildSidecarOutboundTCPTLSFilterChainOpts(node *model.Proxy, env *model.Env
 
 	return out
 }
+
+// Get the most specific matching virtual service for a host. If no virtual service exists, return nil.
+// For example, if virtual services with hosts "*" and "*.com" exist, match "abc.com" to "*.com" instead of "*".
+type virtualServiceForHost func(model.Hostname) *v1alpha3.VirtualService
+
+func makeVirtualServiceForHost(configs []model.Config) virtualServiceForHost {
+	byHostname := make(map[model.Hostname]*v1alpha3.VirtualService)
+	hostnames := make(model.Hostnames, 0)
+	for _, config := range configs {
+		vs := config.Spec.(*v1alpha3.VirtualService)
+		if vs != nil {
+			for _, vsHost := range vs.Hosts {
+				hostname := model.Hostname(vsHost)
+				hostnames = append(hostnames, hostname)
+				byHostname[hostname] = vs
+			}
+		}
+	}
+	sort.Sort(hostnames) // order from most to least specific
+	return func(hostname model.Hostname) *v1alpha3.VirtualService {
+		for _, vsHost := range hostnames {
+			if vsHost.Matches(hostname) {
+				return byHostname[vsHost] // first match is the most specific
+			}
+		}
+		return nil
+	}
+}
```

Comment:
this code can move into the model.. Like model.VirtualServicesForHost() or something.
And while this code is slightly optimized, moving it to model will allow us to optimize this more, by creating that map once in the beginning and using it for all sidecars. See costin's PR for example.